substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,770 @@
+			};
+
+		Delegation::<T>::from(agent, new_delegation_amount).save_or_kill(delegator);
+		ledger.total_delegated =
+			ledger.total_delegated.checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+		ledger.save(agent);
+
+		Self::deposit_event(Event::<T>::Delegated {
+			agent: agent.clone(),
+			delegator: delegator.clone(),
+			amount,
+		});
+
+		Ok(())
+	}
+
+	fn do_release(
+		who: &T::AccountId,
+		delegator: &T::AccountId,
+		amount: BalanceOf<T>,
+		num_slashing_spans: u32,
+	) -> DispatchResult {
+		let mut agent = Agent::<T>::from(who)?;
+		let mut delegation = Delegation::<T>::get(delegator).ok_or(Error::<T>::NotDelegator)?;
+
+		// make sure delegation to be released is sound.
+		ensure!(&delegation.agent == who, Error::<T>::NotAgent);
+		ensure!(delegation.amount >= amount, Error::<T>::NotEnoughFunds);
+
+		// if we do not already have enough funds to be claimed, try withdraw some more.
+		if agent.ledger.unclaimed_withdrawals < amount {
+			// withdraw account.
+			let _ = T::CoreStaking::withdraw_unbonded(who.clone(), num_slashing_spans)
+				.map_err(|_| Error::<T>::WithdrawFailed)?;
+			agent = agent.refresh()?;
+		}
+
+		// if we still do not have enough funds to release, abort.
+		ensure!(agent.ledger.unclaimed_withdrawals >= amount, Error::<T>::NotEnoughFunds);
+
+		// claim withdraw from agent. Kill agent if no delegation left.
+		// TODO(ank4n): Ideally if there is a register, there should be an unregister that should
+		// clean up the agent. Need to improve this.
+		if agent.remove_unclaimed_withdraw(amount)?.save_or_kill()? {
+			let _ = frame_system::Pallet::<T>::dec_providers(who).defensive();
+		}
+		// book keep delegation
+		delegation.amount = delegation
+			.amount
+			.checked_sub(&amount)
+			.defensive_ok_or(ArithmeticError::Overflow)?;
+
+		// remove delegator if nothing delegated anymore
+		delegation.save_or_kill(delegator);
+
+		let released = T::Currency::release(
+			&HoldReason::Delegating.into(),
+			delegator,
+			amount,
+			Precision::BestEffort,
+		)?;
+
+		defensive_assert!(released == amount, "hold should have been released fully");
+
+		Self::deposit_event(Event::<T>::Released {
+			agent: who.clone(),
+			delegator: delegator.clone(),
+			amount,
+		});
+
+		Ok(())
+	}
+
+	/// Migrates delegation of `amount` from `source` account to `destination` account.
+	fn do_migrate_delegation(
+		source_delegator: &T::AccountId,
+		destination_delegator: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		let source_delegation =
+			Delegators::<T>::get(source_delegator).defensive_ok_or(Error::<T>::BadState)?;
+
+		// some checks that must have already been checked before.
+		ensure!(source_delegation.amount >= amount, Error::<T>::NotEnoughFunds);
+		debug_assert!(
+			!Self::is_delegator(destination_delegator) && !Self::is_agent(destination_delegator)
+		);
+
+		// update delegations
+		Delegation::<T>::from(&source_delegation.agent, amount).save_or_kill(destination_delegator);
```

Comment:
Instead of `amount`, should we instead move all the funds held by the proxy for this delegator into the new `destination_delegator`? That would basically mean to remove `amount` as input of this method and just max out the balance based on the delegator's balance in this proxy agent. 

I believe this is better as proxy delegations are temp and we want to clear them out as soon as possible.